url47af6b1535_rolandotr_db_comparison_tgz-pJ8-methodology_HistoryJ8/src/protocols/specifications/TMAProtocol.java

```diff
@@ -0,0 +1,351 @@
+			try {
+				fillDistance2(257);
+			} catch (IOException e) {
+				// TODO Auto-generated catch block
+				e.printStackTrace();
+				throw new RuntimeException(e);
+			}
+		}
+		return distances2[n];
+	}
+	private double[] distances2;
+	/*La informacion sobre el distance esta en un fichero de nombre "d_FDA-"+n*/
+	private void fillDistance2(int n) throws IOException {
+		//System.out.println("loaded file for distance fraud of the New protocol");
+		String tmp;
+		String[] split;
+		int r;
+		distances2 = new double[n];
+		File f = new File(getDistanceFileNameOfValues());
+		BufferedReader reader = new BufferedReader(new FileReader(f));
+		while (reader.ready()){
+			tmp = reader.readLine();
+			split = tmp.split(" ");
+			if (Integer.parseInt(split[0]) == n) break;
+			distances2[Integer.parseInt(split[0])] = Double.parseDouble(split[1]);
+		}
+		reader.close();
+	}	
+
+	
+	private double[] mafia2;
+
+	/** Apr 20, 2010 TnT Comment
+	 * El valor del distance fraud esta en disco. Depende del valor de a, el archivo se llama "tree_a.dat". 
+	 * Y entonces en este archivo podras encontrar la ronda a la que te encuentras.*/
+	private BigDecimal mafiaFraudFromDisk2(int n){
+		if (mafia2 == null) {
+			try {
+				fillMafia2(257);
+			} catch (IOException e) {
+				// TODO Auto-generated catch block
+				e.printStackTrace();
+				throw new RuntimeException(e);
+			}
+		}
+		return new BigDecimal(mafia2[n]);
+	}
+
+	/*La informacion sobre el distance esta en un fichero de nombre "d_FDA-"+n*/
+	private void fillMafia2(int n) throws IOException {
+		//System.out.println("loaded FDA file for mafia fraud");
+		String tmp;
+		String[] split;
+		int r;
+		mafia2 = new double[n];
+		File f = new File(getMafiaFileNameOfValues());
+		BufferedReader reader = new BufferedReader(new FileReader(f));
+		while (reader.ready()){
+			tmp = reader.readLine();
+			split = tmp.split(" ");
+			if (Integer.parseInt(split[0]) == n) break;
+			mafia2[Integer.parseInt(split[0])] = Double.parseDouble(split[1]);
+		}
+		reader.close();
+	}
+	public static void main(String[] args) {
+		TMAProtocol p = new TMAProtocol();
+		p.setNumberOfRounds(256);
+		BigDecimal mafia = p.getMafiaFraudProbabilityExpensiveMethod();
+		System.out.println("n = "+256+" Mafia = "+mafia.doubleValue());
+		BigDecimal distance = p.distanceFraud(p.getNumberOfRounds(), 0);
+		System.out.println("n = "+256+" Distance = "+distance.doubleValue());
+		//generateDataFile();
+	}
+
+	public String getMafiaFileNameOfValues(){
+		return "mafia_"+getAcronym()+"_values.DAT";
+	}
+	public String getDistanceFileNameOfValues(){
+		return "distance_"+getAcronym()+"_values.DAT";
+	}
+
+	public static void generateDataFile() throws IOException{
+		String newLine = System.getProperty("line.separator");
+		TMAProtocol p = new TMAProtocol();
+		FileWriter writerMafia = new FileWriter(new File(p.getMafiaFileNameOfValues()));
+		FileWriter writerDistance = new FileWriter(new File(p.getDistanceFileNameOfValues()));		
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *writerDistance*. The resource is referenced at lines 320, 324. The resource is closed at line 327. There are other execution paths that do not contain closure statements, for example, when OutputStreamWriter.flush throws an exception. Ensure *writerDistance* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).